fetch.js

```diff
@@ -7,451 +7,80 @@
-
-      let postFetchFunctions = [
-        saveFetches(timeStarted, timeEnded, itemsInsertedCount, err, taskResults),
-        saveSources()
-      ];
-
-      if (process.env.SAVE_TO_S3 === 'true' && itemsInsertedCount > 0) {
-        postFetchFunctions.push(saveToS3(itemsInserted));
-      }
-      async.parallel(postFetchFunctions, (err, results) => {
-        if (err) {
-          log.error(err);
-        }
-
-        sendUpdatedWebhook((err) => {
-          if (err) {
-            log.error(err);
-          }
-
-          log.info('Webhook posted, have a good day!');
-          process.exit(0);
-        });
-      });
-    }
-  });
-};
-
-// Branch here depending on whether this is a dryrun or not
-if (argv.dryrun) {
-  log.info('--- Dry run for Testing, nothing is saved to the database. ---');
-  runTasks();
-} else {
-  // Set up DB and add in postgis features
-  pg = knex(knexConfig);
-  st = require('knex-postgis')(pg);
-  log.info('Connected to database.');
-
-  // Run any needed migrations and away we go
-  pg.migrate.latest(knexConfig)
-    .then(() => {
-      log.info('Database migrations are handled, ready to roll!');
-      runTasks();
-    })
-    .catch((e) => {
-      log.error(e);
-    });
-}
+    const fetchReport = {
+      itemsInserted: 0,
+      timeStarted: Date.now(),
+      results: null,
+      errors: null,
+      timeEnded: NaN
+    };
+
+    return DataStream
+      // create a DataStream from sources
+      .fromArray(Object.values(sources))
+      // flatten the sources
+      .flatten()
+      // set parallel limits
+      .setOptions({maxParallel: maxParallelAdapters})
+      // filter sources - if env is set then choose only matching source,
+      //   otherwise filter out inactive sources.
+      // * inactive sources will be run if called by name in env.
+      .use(chooseSourcesBasedOnEnv, env, runningSources)
+      // mark sources as started
+      .do(markSourceAs('started', runningSources))
+      // get measurements object from given source
+      .map(async (source) => getMeasurementsFromSource(source, env))
+      // perform streamed save to DB and S3 on each source.
+      .do(streamMeasurementsToDBAndStorage(doSaveToS3, env, bucketName))
+      // mark sources as finished
+      .do(markSourceAs('finished', runningSources))
+      // handle adapter errors to be forwarded to main stream and well handled.
+      .use(forwardErrors)
+      // convert to measurement report format for storage
+      .map(prepareCompleteResultsMessage(fetchReport, runningSources))
+      // aggregate to Array
+      .toArray()
+      // save fetch log to DB and send a webhook if necessary.
+      .then(
+        // TODO: filter out sources not included in fetch
+        reportAndRecordFetch(fetchReport, sources, env, apiURL, webhookKey)
+      );
+  })()
+])
+  .catch(
+    handleFetchErrors(log, env)
+  )
+  .then(
+    exitCode => process.exit(exitCode || 0)
```

Comment:
Actually no, `handleFetchErrors` extracts the exitCode from the `FetchError` object - if there's something there it will be that (for instance `100`), but if that's not given a default of `0` is used.